### DISCUSSION TITLE
Pinning package versions for `tox` tests?

### DISCUSSION THREAD

**Discussion by brianthelion** [pyscaffold/pyscaffold#488] [Q&A] [answered]
As I understand it, it is *NOT* good practice to "pin" package versions in `setup.cfg`, but instead to leave the version unspecified; pinning should instead be done in a non-canonical artifact like a `requirements.txt`. The `pyscaffold` [docs](https://pyscaffold.org/en/latest/features.html#management-of-requirements-licenses) agree. 

Previously, `pyscaffold` supported `requirements.txt` in most of its workflows. However, more recent versions (v4.0+) have [eliminated](https://github.com/pyscaffold/pyscaffold/issues/182) `requirements.txt` as a standard feature of the package stub.

Vis-a-vis the above, it is currently (v4.0.2) not clear how to pass pinned versions to `tox` for populating testing environments.

Can anyone help there?
Thanks!

### Attempt 1

Added `deps =` to `tox.ini` as follows:

```
[testenv]
description = invoke pytest to run automated tests
isolated_build = True
setenv =
    TOXINIDIR = {toxinidir}
passenv =
    HOME
extras =
    testing
commands =
    pytest {posargs}
deps =
     -rrequirements.txt
```
According to the `tox -vv` logs, this did indeed install the dependencies:

```
[18] /opt/f0cal/my-package/.tox$ /usr/local/bin/python -m virtualenv --no-download --python /usr/local/bin/python default
created virtual environment CPython3.8.11.final.0-64 in 263ms
  creator CPython3Posix(dest=/opt/f0cal/my-package/.tox/default, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/root/.local/share/virtualenv)
    added seed packages: pip==21.2.3, setuptools==57.4.0, wheel==0.37.0
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator
default installdeps: -rrequirements.txt
setting PATH=/opt/f0cal/my-package/.tox/default/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
[31] /opt/f0cal/my-package$ /opt/f0cal/my-package/.tox/default/bin/python -m pip install -i http://localhost:3141/root/dev -rrequirements.txt
Looking in indexes: http://localhost:3141/root/dev
Collecting click<8
  Downloading http://localhost:3141/root/pypi/%2Bf/dac/ca89f4bfadd5d/click-7.1.2-py2.py3-none-any.whl (82 kB)
Collecting Jinja2==2.10.1
  Downloading http://localhost:3141/root/pypi/%2Bf/14d/d6caf1527abb2/Jinja2-2.10.1-py2.py3-none-any.whl (124 kB)
Collecting MarkupSafe>=0.23
  Downloading http://localhost:3141/root/pypi/%2Bf/3dd/007d54ee88b46/MarkupSafe-2.0.1-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (30 kB)
Installing collected packages: MarkupSafe, Jinja2, click
Successfully installed Jinja2-2.10.1 MarkupSafe-2.0.1 click-7.1.2
```
However, the subsequent steps show the the `jinja2==2.10.1` requirement getting clobbered:

```
[46] /opt/f0cal/my-package$ /opt/f0cal/my-package/.tox/default/bin/python -m pip install -i http://localhost:3141/root/dev --exists-action w '/opt/f0cal/my-package/.tox/.tmp/package/1/my-package-0.0.0.zip[testing]'
...
Successfully built my-package
...
Installing collected packages: ,... jinja2, ..., my-package
  Attempting uninstall: jinja2
    Found existing installation: Jinja2 2.10.1
    Uninstalling Jinja2-2.10.1:
      Successfully uninstalled Jinja2-2.10.1
Successfully installed ... jinja2-3.0.1 my-package-0.0.0 ...
```

### Attempt 2

Added `install_command =` to `tox.ini` as follows:

```
[testenv]
description = invoke pytest to run automated tests
isolated_build = True
setenv =
    TOXINIDIR = {toxinidir}
passenv =
    HOME
extras =
    testing
commands =
    pytest {posargs}
install_command =
    pip install -r requirements.txt {opts} {packages}
```

This appears to work correctly.

**Comment by brianthelion** [comment, marked as answer]
See "Attempt 2" in question above.

**Comment by abravalheri** [comment]
Hi @brianthelion sorry for the delay in replying, thanks for the kind words towards the project and for reporting your  solution back.

I am glad to see that you found the answer.

Guaranteeing a deterministic behaviour from `pip install` is a bit complicated, and I suppose `tox` was adding a few flags to the installation steps that allowed the package to be reinstalled. But nice to see that everything worked out with `install_command`.

**Reply by abravalheri** [reply]
The second attempt might also have worked because (I assume) it installs everything using the same pip command (I suppose tox might have been using 2 step installation with `deps`?). That could justify pip being forced to pick the version from `requirements.txt`.
